ze/btx_zeinterval_callbacks.cpp

```diff
@@ -765,87 +765,299 @@ static void zeEventDestroy_exit_callback(void *btx_handle, void *usr_data, int64
  * Sampling
  */
 
-static void lttng_ust_ze_sampling_gpu_energy_callback(void *btx_handle, void *usr_data, int64_t ts,
-                                                      const char *hostname, int64_t vpid,
-                                                      uint64_t vtid, ze_device_handle_t hDevice,
-                                                      uint32_t domain, uint64_t energy,
-                                                      uint64_t sampling_ts) {
-
+std::optional<DeviceHash> get_device_hash(void *usr_data, const char *hostname, int64_t vpid,
+                                          ze_device_handle_t hDevice) {
   auto *data = static_cast<data_t *>(usr_data);
-  auto [it, inserted] = data->device_energy_ref.insert(
-      {{hostname, vpid, hDevice, domain}, {energy, sampling_ts, ts}});
-  // First entry
-  if (inserted)
-    return;
-  auto &[prev_energy, prev_sampling_ts, prev_ts] = it->second;
-
-  // Watt conversion
-  btx_push_message_lttng_power(btx_handle, hostname, 0, 0, prev_ts, BACKEND_ZE, (uint64_t)hDevice,
-                               (thapi_domain_idx)domain,
-                               (energy - prev_energy) / (double)(sampling_ts - prev_sampling_ts));
+  const auto it0 = data->sampling_device_property.find({hostname, vpid, hDevice});
+  if (it0 != data->sampling_device_property.cend()) {
+    const auto &[deviceProp, deviceIdx] = it0->second;
+
+    uint64_t hash = 0xcbf29ce484222325; // FNV offset basis
+    for (int i = 0; i < ZE_MAX_DEVICE_UUID_SIZE; i++) {
+      hash ^= (uint64_t)deviceProp.core.uuid.id[i];
+      hash *= 0x100000001b3; // FNV prime
+    }
 
-  it->second = {energy, sampling_ts, ts};
+    return DeviceHash{hash, deviceIdx}; // Return tuple optional
+  }
+  return std::nullopt; // Return empty optional if not found
 }
 
-static void lttng_ust_ze_sampling_gpu_frequency_callback(void *btx_handle, void *usr_data,
-                                                         int64_t ts, const char *hostname,
-                                                         int64_t vpid, uint64_t vtid,
-                                                         ze_device_handle_t hDevice,
-                                                         uint32_t domain, uint64_t frequency) {
+uint64_t calculate_delta(uint64_t current_val, uint64_t prev_val) {
```

Comment:
Do you need to return `prev_val-current_val` if not equal?  If not then this function is maybe better called `check_delta` or something like that